torch/csrc/jit/passes/to_batch.cpp

```diff
@@ -1,63 +1,441 @@
+    }
+  }
+  r_node->setStage(n->stage());
+  res_graph->appendNode(r_node);
+
+  // create block for Loop node in res_block
+  // if cond is tensor:    first 4 inputs of block: cond_any, cond_data, cond_mask, cond_dims
+  // if cond is not tensor: first 1 input of block: cond
+  auto loop_block = r_node->addBlock();
+  toBatch(n->blocks()[0], loop_block, var_map);
+
+  // change inputs and outputs of block[0] in prim::Loop
+  for(size_t i = EXP_BTENSOR_SIZE - 1; i > 0; i--){
+    loop_block->eraseInput(i);
+  }
+  if(cond_is_tensor){
+    for(size_t i = 0; i < EXP_BTENSOR_SIZE; i++){
+      loop_block->insertInput(i + 1, "cond_" + EXP_BTENSOR_NAME[i]);
+    }
+  }
+
+  WithInsertPoint guard(loop_block);
+
+  // use where operator to update variables and add to outputs
+  if(cond_is_tensor){
+    for(size_t i = 0; i < n->outputs().size(); i++){
+      std::vector<Value*> inputs;
+      for(size_t j = 0; j < EXP_BTENSOR_SIZE; j++){
+        inputs.push_back(loop_block->inputs()[j + 1]);
+      }
+      auto data = batch_map.at(n->blocks()[0]->outputs()[i + 1]);
+      inputs.insert(inputs.end(), data.begin(), data.end());
+      for(size_t j = 0; j < EXP_BTENSOR_SIZE; j++){
+        inputs.push_back(loop_block->inputs()[i * EXP_BTENSOR_SIZE + j + EXP_BTENSOR_SIZE + 1]);
+      }
+      auto outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_operator_table.at("where"), inputs);
+      batch_map[n->outputs()[i]] = outputs;
+      for(size_t j = 0; j < EXP_BTENSOR_SIZE; j++){
+        loop_block->registerOutput(outputs[j]);
+      }
+    }
+  }
+  // add outputs
+  else{
+    for(size_t i = 0; i < n->outputs().size(); i++){
+      auto outputs = batch_map.at(n->blocks()[0]->outputs()[i + 1]);
+      for(size_t j = 0; j < EXP_BTENSOR_SIZE; j++){
+        loop_block->registerOutput(outputs[j]);
+      }
+    }
+  }
+
+  // update loop conditions
+  if(cond_is_tensor){
+    auto cond = batch_map.at(n->blocks()[0]->outputs()[0]);
+    auto cond_any = script::inlineCallTo(*res_block->owningGraph(), *batch_operator_table.at("any"), cond);
+    loop_block->insertOutput(0, cond_any[0]);
+    for(size_t i = 0; i < EXP_BTENSOR_SIZE; i++){
+      loop_block->insertOutput(i + 1, cond[i]);
+    }
+  }
+  else{
+    auto cond = rn_env.at(n->blocks()[0]->outputs()[0]);
+    loop_block->insertOutput(0, cond);
+  }
+
+  // change outputs of prim::Loop
+  auto size = r_node->outputs().size();
+  for(size_t i = 0; i < size; i++){
+    for(size_t j = 1; j < EXP_BTENSOR_SIZE; j++){
+      r_node->insertOutput(i * EXP_BTENSOR_SIZE + j);
+    }
+    batch_map[n->outputs()[i]] = r_node->outputs().slice(i * EXP_BTENSOR_SIZE, EXP_BTENSOR_SIZE);
+  }
+  // add cond to outputs of loop node
+  if(cond_is_tensor){
+    for(size_t i = 0; i < EXP_BTENSOR_SIZE; i++){
+      r_node->insertOutput(i);
+    }
+  }
+}
+
+void ToBatch::toBatch(Block* block, Block* res_block, std::unordered_map<std::string, Value*>& upper_var_map) {
```

Comment:
The string names for Value nodes should be used for _debugging only_. They can be set to absolutely any string by an optimization pass. If you are putting them in a map as a key, this is a strong indication that there is a problem with how the pass is arranged. I suggest trying to change this to map from Value* to Value*, which is much more likely what you want.